flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala

```diff
@@ -152,21 +153,40 @@ abstract class TableEnvironment(val config: TableConfig) {
   protected def getBuiltInRuleSet: RuleSet
 
   /**
-    * Registers a [[UserDefinedFunction]] under a unique name. Replaces already existing
+    * Registers a [[ScalarFunction]] under a unique name. Replaces already existing
     * user-defined functions under this name.
     */
-  def registerFunction(name: String, function: UserDefinedFunction): Unit = {
-    function match {
-      case sf: ScalarFunction =>
-        // register in Table API
-        functionCatalog.registerFunction(name, function.getClass)
+  def registerFunction(name: String, function: ScalarFunction): Unit = {
+    // register in Table API
+    functionCatalog.registerFunction(name, function.getClass)
 
-        // register in SQL API
-        functionCatalog.registerSqlFunction(sf.getSqlFunction(name, typeFactory))
+    // register in SQL API
+    functionCatalog.registerSqlFunction(function.getSqlFunction(name, typeFactory))
+  }
+
+  /**
+    * Registers a [[TableFunction]] under a unique name. Replaces already existing
+    * user-defined functions under this name.
+    */
+  private[flink] def registerTableFunctionInternal[T: TypeInformation](
+    name: String, tf: TableFunction[T]): Unit = {
 
-      case _ =>
-        throw new TableException("Unsupported user-defined function type.")
+    val typeInfo: TypeInformation[_] = if (tf.getResultType != null) {
+      tf.getResultType
+    } else {
+      implicitly[TypeInformation[T]]
     }
+
+    val (fieldNames, fieldIndexes) = UserDefinedFunctionUtils.getFieldInfo(typeInfo)
```

Comment:
It is possible. But I would like the move `createSqlFunction`  and `createSqlFunctions` to utils.